lib/driver-cli.js

```diff
@@ -0,0 +1,265 @@
+         INSTALL_TYPE_LOCAL, INSTALL_TYPE_NPM, INSTALL_TYPE_GIT,
+         INSTALL_TYPE_GITHUB } from './drivers';
+import { getLatestVersion, installPackage, linkPackage } from './npm';
+import semver from 'semver';
+
+let INSTALLED_DRIVERS;
+
+function errAndQuit (json, msg) {
+  if (json) {
+    console.log(JSON.stringify({error: msg}));
+  } else {
+    console.error(msg.red);
+  }
+  process.exit(1);
+}
+
+function log (json, msg) {
+  !json && console.log(msg);
+}
+
+async function runDriverCommand (args) {
+  let jsonResult = null;
+  try {
+    INSTALLED_DRIVERS = validateDriversJson();
+    switch (args.driverCommand) {
+      case 'list':
+        jsonResult = await list(args);
+        break;
+      case 'install':
+        jsonResult = await install(args);
+        break;
+      case 'uninstall':
+        jsonResult = await uninstall(args);
+        break;
+      case 'update':
+        jsonResult = await update(args);
+        break;
+      default:
+        throw new Error(`Cannot handle driver command ${args.driverCommand}`);
+    }
+  } catch (err) {
+    errAndQuit(args.json, err.message);
+  }
+
+  if (args.json) {
+    console.log(JSON.stringify(jsonResult));
+  }
+}
+
+function startSpinner (msg) {
+  process.stdout.write(msg);
+
+  let count = 1;
+  const numDots = 5;
+  const dotTime = 500;
+  const cursorStart = msg.length;
+
+  const interval = setInterval(() => {
+    if (count % (numDots + 1) === 0) {
+      process.stdout.cursorTo(cursorStart);
+      for (let i = 0; i < numDots; i++) {
+        process.stdout.write(' ');
+      }
+      process.stdout.cursorTo(cursorStart);
+    } else {
+      process.stdout.write('.');
+    }
+    count++;
+  }, dotTime);
+
+  return () => {
+    clearInterval(interval);
+    process.stdout.write('\n');
+  };
+}
+
+function logSpin (json, msg) {
+  if (json) {
+    return () => {};
+  }
+
+  return startSpinner(msg);
+}
+
+async function list ({json, showInstalled, showUpdates}) {
+  const lsMsg = `Listing ${showInstalled ? 'installed' : 'available'} drivers`;
+  const stopSpinner = logSpin(json, lsMsg);
+
+  const drivers = {};
+
+  for (const name of Object.keys({...INSTALLED_DRIVERS})) {
+    drivers[name] = INSTALLED_DRIVERS[name];
+    drivers[name].installed = true;
+  }
+
+  if (!showInstalled) {
+    for (const name of Object.keys(KNOWN_DRIVERS)) {
+      if (!drivers[name]) {
+        drivers[name] = {
+          pkgName: KNOWN_DRIVERS[name],
+          installed: false,
+        };
+      }
+    }
+  }
+
+  if (showUpdates) {
+    for (const name of Object.keys(drivers)) {
+      const {version, pkgName, installType} = drivers[name];
+      drivers[name].updateVersion = null;
+      if (installType === INSTALL_TYPE_NPM) {
+        const latestVersion = await getLatestVersion(pkgName);
+        if (!version || semver.gt(latestVersion, version)) {
+          drivers[name].updateVersion = latestVersion;
+        }
+      }
+    }
+  }
+
+  stopSpinner();
```

Comment:
should this be called in finally block?

Reply:
ah yes, good point.

Reply:
in practice, every error thrown ends up halting the process, so i think it's fine. but finally is much safer/cleaner.